spinnaker-monitoring-daemon/spinnaker-monitoring/spectator_metric_transformer.py

```diff
@@ -25,15 +25,24 @@
 import sys
 import yaml
 
-
-# see https://stackoverflow.com/questions/1175208/\
-#     elegant-python-function-to-convert-camelcase-to-snake-case
-_SNAKE_CASE_RE = re.compile('((?<=[a-z0-9])[A-Z]|(?!^)(?<!_)[A-Z](?=[a-z]))')
-
+# This is a cache for the _snakeify function to increase performance on the
+#    conversion of CamelCase to snake_case
+SNAKEIFY_CACHE = {}
```

Comment:
I'm curious how much this cache helps vs. the risk of having a memory leak here.  Given that there's no expiration here, this will grow without bound; while it's only holding strings I do still worry about having this here.  If we need to memoize, is there maybe a library we can use that limits the size of the cache?

Reply:
When I ran the unittest 1,000,000 times, the regex took about 44 seconds, the function without cache took about 28 seconds, and the cache took about 2.5 seconds.  I'm also not as concerned with the cache growing too much as the names and labels that this function is being used for is from the spectator monitoring endpoint, which should have a finite count and be repeated every 30 or so seconds when pulling the mentrics to (in our clients case) prometheus.

Reply:
I think this is a case where the performance of repeating the unit test might not be a great proxy for what would happen in production...repeating the test with the same keys would mean a 99.9999% hit rate of the cache.

Obviously if that's similar to the cache hit rate in production then this would be a good proxy...do you have an idea for the typical number of keys that will be here, and how often they change?  If this is a static list of keys that don't change, this is probably reasonable, but just want to make sure we check that assumption independently.

Reply:
I think this is a case where the performance of repeating the unit test might not be a great proxy for what would happen in production...repeating the test with the same keys would mean a 99.9999% hit rate of the cache.

Obviously if that's similar to the cache hit rate in production then this would be a good proxy...do you have an idea for the typical number of keys that will be here, and how often they change?  If this is a static list of keys that don't change, this is probably reasonable, but just want to make sure we check that assumption independently.

Reply:
This is the only area in the code that the `_snakeify` function is used: https://github.com/spinnaker/spinnaker-monitoring/blob/a724ee721df36d347530313e36c0dab241126c67/spinnaker-monitoring-daemon/spinnaker-monitoring/spectator_metric_transformer.py#L929
It appears to only be used to convert metric labels and metric names, which should be coming from the code in the Spinnaker service being monitored (e.g. clouddriver).  If metric labels and names are constantly changing, then monitoring would not be effective.  Please verify this, but as far as I can tell, this will be a set list of labels and name that will be created initially, on the first request or two, then come from the cache table afterwards for the rest of the time the daemon is running.